### PR TITLE
limp_nopred option for more tolerable play on high ping

### PR DESCRIPTION
**Pull request by Iceman12k**
New client option which makes use of PMF_NO_PREDICTION while legs are damaged. This prevents the janky warping which can get especially bad on high ping.

I'm new to Q2 modding (but very accustomed to QW modding) so I may have missed a step to initialize the cvar on the client side if that is even a thing. Just a bit clunky that the client would have to do `set limp_nopred 1 u` to initialize it as a userinfo cvar.

### CODE REVIEW

**Comment by Raptor007** [comment]
I like it!  Anything to help mitigate the problems with high ping is worth trying, and I appreciate that the clients can choose whether to enable this.  I should check its effects before pulling into this master branch.

If this works really well, it may be worth incorporating a new server cvar to specify a ping level that would automatically enable this effect for laggy clients.

Might make sense to move the flag changes from the damage and bandage functions into ClientThink where it considers leg damage:
https://github.com/aq2-tng/aq2-tng/blob/master/source/p_client.c#L2944-L2960

**Comment by Iceman12k** [comment]
The only reason I didn't put it there is potential conflicts with chasecam and grapple hook (tho I suspect that code is just leftover baseq2 stuff, never remember seeing anyone spawn a hook in AQ2 :P)

**Comment by hifi** [comment]
> The only reason I didn't put it there is potential conflicts with chasecam and grapple hook (tho I suspect that code is just leftover baseq2 stuff, never remember seeing anyone spawn a hook in AQ2 :P)

I added grapple hook to TNG CTF years ago and we played with it enabled quite a lot. It makes CTF much faster and easier to score in large maps.

**Comment by Iceman12k** [comment]
Alright, added some auto-activation behavior for clients who don't have the userinfo set. `sv_limp_highping` is the cvar name, feel free to change it to fit some convention. I'm not great at naming cvars.

It has some simple protection against swingy-ping causing weird behavior, and I stuck the whole thing in a ClientLegDamage function to prevent copy-pasted code. Honestly the whole `client->leghits++` and `client->leg_damage = 1` stuff could also be stuck in there but I didn't want to touch unrelated code.

**Comment by Raptor007** [comment]
Would this placement mess with chasecam or grapple?  It seems to me it wouldn't, and might be simpler to have in one place rather than changing code wherever legs may take or heal damage:
```
		if( client->leg_damage && ent->groundentity && ! has_enhanced_slippers )
		{
			int frame_mod_6 = (level.framenum / game.framediv) % 6;
			if( frame_mod_6 <= 2 )
			{
				pm.cmd.forwardmove = 0;
				pm.cmd.sidemove = 0;
			}
			else if( frame_mod_6 == 3 )
			{
				pm.cmd.forwardmove /= client->leghits + 1;
				pm.cmd.sidemove /= client->leghits + 1;
			}

			// Prevent jumping with leg damage.
			pm.s.pm_flags |= PMF_JUMP_HELD;

			// Set prediction bit for limp_nopred.
			ClientLegDamage(ent);
		}
		else
			ent->client->ps.pmove.pm_flags &= ~PMF_NO_PREDICTION;
```

**Comment by Iceman12k** [comment]
I think it would turn prediction back on for people using a hook, and not disable prediction for leg damage until a player touches the ground. Chasecam would be fine I think. Does bring up a good point that enhanced slippers should have prediction stay on tho.

**Comment by Raptor007** [comment]
Ah I see what you mean about grapple.  This might solve it: `else if( ent->client->ctf_grapplestate <= CTF_GRAPPLE_STATE_FLY )`

**Comment by Iceman12k** [comment]
I just opted to move the leg damage stuff into their respective functions, so that way code isn't copy-pasted in 2-3 different locations. Since the prediction-disabling logic is supposed to run only once per leg damage, I'd have to add an additional entity field to keep track of it if I was running it per client input frame in `ClientThink`. It's like that so ping fluctuations at least maintain their prediction setting until next time you break your legs as it would be disorienting for prediction to enable/disable multiple times in a short period.

**Comment by Raptor007** [comment]
Finally got a little time to pull this to my machine and try it rather than just looking at the code and guessing.

Unfortunately `setu` does not trigger a call to `ClientUserinfoChanged` when it first encounters a custom variable name.  I had to either `setu limp_nopred 1` twice for it to work, follow it with another userinfo change like name, or reconnect.  We could just say that's fine as long as we know about it, or this variable could be handled in g_cmds.c by adding `{ "limp_nopred", Cmd_LimpNoPred_f, 0 },` to commandList and moving/tweaking your logic from the bottom of `ClientUserinfoChanged` into a new `Cmd_LimpNoPred_f` function.  This would have the advantage (or disadvantage?) of not retaining the setting across server changes.

Because the effect is applied upon leg damage or heal rather than during `ClientThink`, changing `limp_nopred` while limping doesn't change the prediction flag.  Players deciding which mode they prefer would likely want to be able to toggle it while limping, as I was attempting to do while testing.  Instead I had to go jump off a ledge again to compare.  Again, not a big deal because aside from some initial experimentation, once a player sets it they'll probably leave it alone.

One more stupid nitpick, this `case -1:` will never match because for type int, -1 & 255 is 255, but it really doesn't matter because there's no default case you're trying to avoid anyway.
https://github.com/aq2-tng/aq2-tng/pull/86/files#diff-48891edec6a9ade15b5bf1991e6ed3f67e8257baee56a1d4734c9e2473430668R2194-R2197

All that said, none of these are big problems and I would be happy to merge this now if you're ready for me to do that.

Also, now that you've dragged the old leg damage code into the light, I see that `leg_dam_count` and `leg_sound` have entirely no effect, and even `leg_damage` is redundant because `leghits` could just be checked instead.  I won't hold up your merge over this, just noting it for myself to consider refactoring later.

**Review by Raptor007 — APPROVED** [review]

**Pull request merged**
